### PR TITLE
AP-9902 # Fixed checkForPaymentEvent not throwing if calculation is missing elements

### PR DESCRIPTION
**Pull request by mymattcarroll**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [ ] Have you tested your implementation locally?
- [ ] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

---

> [!NOTE]
> **Medium Risk**
> Changes payment submission behavior and makes `evaluateExpression` a breaking API for any external callers; fixes are in payment-critical configuration paths with expanded test coverage.
> 
> **Overview**
> **`evaluateExpression`** no longer returns `number | undefined` or throws on empty/parse failures. It returns **`EvaluateExpressionResult`**: `{ type: 'RESULT', value }`, `{ type: 'MISSING_VALUES' }`, or `{ type: 'INVALID_EXPRESSION', error }` (unexpected runtime errors still throw).
> 
> Adds **`findMissingFormElementsInExpression`** to detect `{ELEMENT:...}` paths that are not on the form definition (including nested **`Items|Amount`** vs root **`Amount`** for repeatable sets and nested forms). Intended for server-side payment validation, not interactive UI.
> 
> **`checkForPaymentEvent`** for **`EXPRESSION`** amounts now runs that check before evaluation: missing definition paths throw **`[PAYMENT_EXPRESSION_MISSING_FORM_ELEMENTS]`** (fixes cases that previously looked like missing submission data and skipped payment). Invalid expressions throw **`[PAYMENT_INVALID_EXPRESSION]`**; missing submission values still skip payment. Other payment configuration errors gain **`[PAYMENT_*]`** error-code prefixes.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 05e553d80debecd39dcd7748600e4f5db36eb2da. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/paymentService.ts [hunk 1]
```diff
@@ -206,11 +247,15 @@ function resolvePaymentEventAmount({
...
-      return getRootElementValueById(
+      const amount = getRootElementValueById(
         amountElement.id,
         definition.elements,
         submission,
       )
+
+      if (typeof amount === 'number' && !Number.isNaN(amount)) {
+        return amount
+      }
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/paymentService.ts]: Fixed.

File: src/calculationService.ts [hunk 2]
```diff
@@ -36,6 +41,76 @@ function roundToFixed(number: number, decimals: number) {
...
+  const parts = elementName.split('|')
+  let searchElements = formElements
+  let found: FormTypes.FormElement | undefined
+
+  for (let i = 0; i < parts.length; i++) {
+    const name = parts[i]
+    found = findFormElement(
+      searchElements,
+      (element) => 'name' in element && element.name === name,
+    )
```
[the remark this answers is not in the document]
Comment (mymattcarroll) [reply src/calculationService.ts:59]: Fixed

**Review by divporter — APPROVED** [review]

**Comment by divporter** [comment]
Does apps-react swallow the errors?

**Comment by mymattcarroll** [comment]
> Does apps-react swallow the errors?

The calculation form element will swallow the errors and show generic message at the moment.

Under normal circumstances (assuming the calc path is valid) none of the errors I prefixed should be shown in the "check for payment" stuff. It will only show error messages if something was misconfigured or user is trying to be naughty.
